D8 Create and Build

generalredneck opened the pull request:
Built off of @patrickocoffeyo's D8 branch and translated some of the changes to the new standard. Also, Fixed up the src files added by Patrick to actually work with D8.
## Testing

To Test, you can build a new d8 site by doing `aquifer projectname -d 8` then do an `aquifer build` inside the project.

If you haven't tested a stand alone version of aquifer before... do the following:
- clone aquifer to a directory `git clone https://github.com/aquifer/aquifer.git`
- add my fork as a remote `git remote add generalredneck https://github.com/generalredneck/aquifer.git`
- check out the **d8** branch. `git fetch generalredneck; git checkout generalredneck/d8`
- run `npm install`
- navigate to where you want to build a project.
- run `/PATH_TO_AQUIFER_CLONE/bin/aquifer.js projectname -d 8`
- `cd projectname`
- `/PATH_TO_AQUIFER_CLONE/bin/aquifer.js build`

**Expected Results**:
You should have a fully built out d8 site that if you set up `settings/secret.settings.php` with a database and run from a web server, should have a fully operational d8 site. The standard module, themes, and root folders should point to the respective points in d8.

patrickocoffeyo committed:
Move default files into d7 and d8 folders

patrickocoffeyo committed:
Make src files depend on major drupal version, and bake into project api

Commit:
Merge remote-tracking branch 'origin/D8' into d8

Commit:
Updating Patricks code for the create command to work with the 1.0.0 rewrite.

Commit:
Fixing the drush commands.

Commit:
Removing Features Extra because it doesn't exist for d8.

Commit:
Fixing Build for d8.

patrickocoffeyo commented on lib/build.api.js:
```diff
@@ -136,23 +153,27 @@ class Build {
-          dest: 'sites/all/themes/custom',
-          type: 'dir'
+        // Add themes links.
+        Object.keys(project.config.paths.themes).forEach((key) => {
```
As far as I know there is only one custom theme directory... does this need to be in a loop?

generalredneck replied:
What this does though is kinda simplify the "map" above. Notice that the themes directory is the only one that specifically calls for "custom". That being said, it was just "wierd" to me to have a map that was "themesCustom"

patrickocoffeyo replied:
Ah, that makes sense. Thanks!

patrickocoffeyo commented on lib/project.api.js:
```diff
@@ -23,15 +23,17 @@ class Project {
     this.paths = {};
     this.absolutePaths = {};
-    this.srcDir = path.join(path.dirname(fs.realpathSync(__filename)), '../src');
+    this.drupalVersion = typeof drupalVersion === 'undefined' ? 7 : drupalVersion;
```
``` javascript
this.drupalVersion = drupalVersion || 7;
```

generalredneck replied:
Can tell I'm a PHP dev  :P

patrickocoffeyo commented on src/d8/default.settings.php:
```diff
@@ -0,0 +1,92 @@
+ */
+$conf['install_profile'] = 'standard';
+$conf['master_version'] = 2;
+$conf['master_modules'] = array(
```
We need to go through this list for D8 and make sure that we a) disable the right core modules and b) enable the right custom modules. For instance, we don't download Fences for D8 (yay twig!)

patrickocoffeyo commented on src/d8/default.settings.php:
```diff
@@ -0,0 +1,92 @@
+
+$conf['404_fast_paths_exclude'] = '/\/(?:styles)\//';
+$conf['404_fast_paths'] = '/\.(?:txt|png|gif|jpe?g|css|js|ico|swf|flv|cgi|bat|pl|dll|exe|asp)$/i';
+$conf['404_fast_html'] = '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML+RDFa 1.0//EN" "http://www.w3.org/MarkUp/DTD/xhtml-rdfa-1.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>404 Not Found</title></head><body><h1>Not Found</h1><p>The requested URL "@path" was not found on this server.</p></body></html>';
```
Is the fast 404 stuff the same in D8?

patrickocoffeyo commented on src/d8/drupal.make.yml:
```diff
@@ -0,0 +1,21 @@
+  pathauto:
+    version: ~
+  token:
+    version: ~
```
We should make sure this list of default modules is reasonable.

thebruce replied:
some of these definitely are not ready today, culling recommended. will see if i can help

generalredneck replied:
that's been done actually. if anything we might want to add some to it. I did weed out ones that weren't availiable like master, features_extra, and the aurora theme.

patrickocoffeyo replied:
So, not having the master module is a pretty big deal for Aquifer (see settings.php file) and it's a big part of how Aquifer works. Before we merge this in and call it done, we will probably need to port the Master module to D8.

patrickocoffeyo commented on lib/build.api.js:
```diff
@@ -110,13 +110,30 @@ class Build {
+            'site': 'sites/default'
+          }
+        };
+        let versionDestPaths = destPaths[project.drupalVersion];
```
This variable never changes after being set, let's use `const` instead of `let`

patrickocoffeyo replied:
I think this should be a class property, `this.paths.destinations` or something, simply because extensions will need to be able to interact with files without having to specify the Drupal version.

Commit:
Updating the default d8 settings files.

patrickocoffeyo commented:
This seems to work great! Lets make 100% sure that robots.txt, .htaccess, and settings files are D8 compatible, in my original branch I literally just made a copy of d7 -> d8 and didn't get around to doing that.

Before we merge, I'd like to get a version of this running in a DrupalVM instance and make sure refresh works, and all the assets are correctly linked.

patrickocoffeyo commented:
Let's also add some documentation around this new feature in the README.md file, just the new flag :) Also, lets get you added to the contributors list in the README!!

Commit:
Adding core.extension.yml to the mix.

Commit:
Making suggested changes by patrickocoffeyo

Commit:
Updating readme.

Revert "Adding core.extension.yml to the mix."

This reverts commit 380d3e038135a09c3cf16ff6a2ba1d8041c17406 since this is not a valid way to work around master.

Fixing Default refresh and drush make.

generalredneck commented:
@patrickocoffeyo, I've worked out your comments. Could you take another look? I would change the labels to "Needs Review" but I can't add to it.

patrickocoffeyo commented on README.md:
```diff
@@ -1,21 +1,32 @@
-![Aquifer](https://raw.githubusercontent.com/aquifer/aquifer.io/master/branding/aquifer-logo-drupal-1456.png)
+![Aquifer][logo]
```
What is the benefit behind this change?

generalredneck replied:
Basically I reformatted the readme to be more viewable by terminal editors and hard wraps at 80 chars. By using the alternate version of links and images, you can place their URLS in a list at the bottom of a section making them more organized and allowing the text itself to wrap easier. See [Link Markdown documentation](https://daringfireball.net/projects/markdown/syntax#link)

patrickocoffeyo replied:
OK, cool, I can get behind that! I like it. 

> Basically I reformatted the readme to be more viewable by terminal editors

I'm not too sure about this, as someone that uses a terminal editor exclusively, I'd rather set my own width and wrapping rules.

patrickocoffeyo commented on README.md:
```diff
@@ -1,21 +1,32 @@
+![Aquifer][logo]
 
-Aquifer is a command line interface that makes it easy to scaffold, build, test, and deploy your Drupal websites. It provides a default set of tools that allow you to develop, and build Drupal sites using the Drush-make workflow. In addition, Aquifer ships with an extensions system that allows you to add additional tools to your project.
+Aquifer is a command line interface that makes it easy to scaffold, build, test,
```
Can we omit all manual wrapping? The only place I've ever seen that is on d.o :D

generalredneck replied:
See comment on image.

patrickocoffeyo commented on README.md:
```diff
@@ -1,21 +1,32 @@
 
 Aquifer should now be installed!
 
+[i1]: https://github.com/creationix/nvm
```
bump.. what's the benefit of saving paths as variables?

generalredneck replied:
See comment on image.

patrickocoffeyo commented on lib/build.api.js:
```diff
@@ -110,13 +110,30 @@ class Build {
+            'site': 'sites/default'
+          }
+        };
+        const versionDestPaths = destPaths[project.drupalVersion];
```
I think this needs to be a class property. That way, when extensions can easily support d7/8 without having to create their own drupal version specific path maps.

Commit:
Fixing patrickocoffeyo's suggestion on destPaths:

> I think this needs to be a class property. That way, when extensions can easily support d7/8 without having to create their own drupal version specific path maps.

Commit:
Updating d8 default files.

generalredneck commented:
Ready for another look

patrickocoffeyo commented:
@generalredneck This looks great and works as described. Great work, let's get this merged into 1.0.0!
